### ISSUE TITLE
The (if _ then _ else) pattern is not recognized when using ifN on a simple goal

### ISSUE THREAD

**Issue by pochinha** [math-comp/math-comp#549]
I'm not sure what the underlying problem is.

Example:

```
From mathcomp Require Import all_ssreflect.

Goal (if 1 < 0 then true else false) = false.
Proof.
  Fail rewrite ifN. (* The LHS of ifN (if _ then _ else _) does not match any subterm of the goal *)
  Fail apply: ifN. (* Cannot apply lemma ifN *)
  apply ifN (* works *)
Abort.
```

Using `rewrite (@ifN _ (1 < 0)).` also works.

Happens both with Coq 8.9.1 + mathcomp 1.9.0 and with Coq 8.11.2 + mathcomp 1.11.0, which are the versions I have installed.

**Comment by chdoc** [comment]
I can confirm this behavior in coq-8.12+mathcomp-1.11.0

> Using `rewrite (@ifN _ (1 < 0)).` also works.
> EDIT: using `rewrite (@ifN _ (1 < 0)).` leaves a dangling `bool`. Furthermore, `rewrite-> ifN` works properly.

This actually hints at what's going on. This rewrite is not actually rewriting the if expression. Indeed, one can do:
```coq
Goal (if 1 < 0 then true else false) = false.
have b : bool by exact false.
have E := @ifN _ (1 < 0) b false.
rewrite E. all:reflexivity.
Qed.
```
The rewrite that is actually performed is rewriting the reducts `false = false`. This manages to even bypass the progress checker. :fearful: 

The problem seems to only occur if the `if _ then _ else _` pattern can reduce, because the rewrite succeeds when blocking the reduction of the if-expression beforehand:
```coq
Goal (if 1 < 0 then true else false) = false.
by rewrite [1 < 0]lock ifN // -lock.
Qed.
```

**Comment by CohenCyril** [comment]
@chdoc cf https://github.com/coq/coq/issues/12780

**Comment by chdoc** [comment]
> @chdoc cf [coq/coq#12780](https://github.com/coq/coq/issues/12780)

Yes, I had seen that issue, and I agree that it is probably related. However, since in this instance `rewrite -> ifN` succeeds while `rewrite ifN` does not, there seems to be an issue with the way the ssreflect rewrite matches if-expressions, which should be distinct from unification. 

I also wanted to understand and document how the `bool` subgoal arises.

In any event, I suppose this should be in the issue tracker of coq, as it doesn't concern mathcomp.

**Comment by CohenCyril** [comment]
> Yes, I had seen that issue, and I agree that it is probably related. However, since in this instance `rewrite -> ifN` succeeds while `rewrite ifN` does not, there seems to be an issue with the way the ssreflect rewrite matches if-expressions, which should be distinct from unification.

I see, no that's just because the two rewrites do not rely on the same unification algorithms... ssr rewrite relies on evarconv.ml while coq's standard rewrite relies on unification.ml
